### DISCUSSION TITLE
New Update to Dev Branch - Modules, Menus, Buttons, Hopper Sensors!

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#22] [General]
[Module Selection, Menu / Physical Button Inputs, Hopper Sensors Added](https://github.com/nebhead/PiFire/commit/e8a48d5698bd772abe4f2453f792fad9d221e441)

Some large changes to this commit.  First, a module selection script has been added to PiFire which means that it is easier to select different modules to be used with PiFire.  This has also been integrated into the setup/install script to install the right libraries for the current system.  

Launch the module selection at the command prompt by typing: 
`$ sudo bash modules.sh `

The module selection will also affect how the flask app behaves.  If the prototype module is selected as the grill platform, it assumes you are running on a PC and launches in debug mode. 

Any module configuration changes will require a restart (either via the supervisord or via a reboot).

**Other changes:** 
- Menu / Button Inputs:  Physical button inputs have been added to the SSD1306 display, which requires you to select the new SSD1306b display module.  A basic menu will pop up when any of the buttons is pressed and allows you to take basic actions from this menu.  It's rather crudely implemented, but it seems to work.  Needs more polish. 
- Hopper Level Sensor Modules:  Added modules to support hopper level sensing with the vl53l0x Time of Flight sensor and the HC-SR04 Ultrasonic Sensor (untested).  When using either of these modules, the dashboard will update with the hopper pellet level, as well as on status on the pellet manager page.  These pellet levels will update periodically (every 2.5m on the dashboard and every 1m on the pellet manager).  **Note:** The VL530L0X has been tested, but is _definitely buggy_.  The HC-SR04 has not been tested on real hardware.  

**Note:** 
- If you are updating from an existing install, please delete your settings.json and control.json files prior to running. 
- Reminder if installing from scratch, remember to use raspi-config to enable I2C for any devices that may utilize this bus (i.e. ADC, OLED)

**Comment by faisalhusain007** [comment]
I will test and report back. Thank you for the updates.

**Comment by faisalhusain007** [comment]
The buttons seem to work, they do the job, I think. Just out of interest are you using any pullups on the buttons?

The TOF sensor I see what you mean about the fluctuations, I tried a different libray and adapted the code slightly but similar results. I then tried a different sensor That works much better more stable and accurate. Here are the sensors I used :-
![WhatsApp Image 2021-01-22 at 14 31 44](https://user-images.githubusercontent.com/36872476/105458010-c1d05980-5cc2-11eb-85a8-65456c8779fb.jpeg)
 This sensor above is not very stable very erratic.

However this sensor bellow using the libs you suggest is very stable on my bench and works great so far
![WhatsApp Image 2021-01-22 at 14 32 40](https://user-images.githubusercontent.com/36872476/105458229-1f64a600-5cc3-11eb-8c68-c00a0a0f4945.jpeg)

According to the datasheets they both appear to be using the same chip, so I am not sure what the difference is other than the PCB foot print.

I will investigate further and see what I find.

Just as a quick request would it be possible to show the distance detected by the tof for testing purposes?

**Reply by faisalhusain007** [reply]
I have been playing around a lot with these sensors, what is apparent is they appear to be sensitive to even the slightest change in angles, I will print a holder for the TOF sensor and test further, but so far it does appear to work. On my bench I use tape to hold down the sensor, as the tape gives a little I get weird readings, this is clear when I check the distance being read by the sensor, I figured out as it moves slightly the readings change drastically the further away the target is picking up something closer. I will try and test again with the sensor in a printed enclosure.

**Reply by nebhead** [reply]
> The buttons seem to work, they do the job, I think. Just out of interest are you using any pullups on the buttons?

Only using the internal pull-ups.  I figured they would be enough to do the trick, then use a 330Ohm resistor to ground.  I have had a few phantom button presses, but I think that's just my terrible soldering that caused it.  

> However this sensor bellow using the libs you suggest is very stable on my bench and works great so far

This is what I'm using:

![image](https://user-images.githubusercontent.com/19876886/105520496-a3049f80-5c8f-11eb-81bd-b4a64243d2c4.png)

Mine looks very similar to your second sensor.  The problem I see is that it will take readings initially, after a reboot.  And the readings are reasonably consistent and accurate.  However, after a while something happens and I just start getting bad readings consistently.  This is why I thought I might switch to the HC-SR04 sensor and see if that is more consistent.  

> According to the datasheets they both appear to be using the same chip, so I am not sure what the difference is other than the PCB foot print.

I wonder if there is blockage on the laser, or the receiver that causes some erroneous readings maybe?  My confidence level with these is not super high.  

> Just as a quick request would it be possible to show the distance detected by the tof for testing purposes?

Yep.  Easiest way to do this is to add a print statement in `distance_vl53l0x.py` and then run the `control.py` script from the terminal. 

```
		# Convert mm to cm 
		AvgDist = AvgDist / 10 
		print(f'Average Distance Measured: {AvgDist}cm')  # <=== Added this line
		
		# If Average Distance is less than the empty distance, calculate percentage
```

First, stop the current `control.py` script from running from the Supervisor WebUI, then run the script manually from the terminal:

`sudo python3 control.py`

Another option might be to import the DebugWrite function from the common.py, and write out to the debug.  Actually, now that I'm thinking about it, that might be a better option.  You can do this by modifying the following in `distance_vl53l0x.py`: 

In the import section, import the DebugWrite function:
```
import VL53L0X
import time 
from common import DebugWrite # <=== Insert this line
```
Then further down, after the AvgDist is calculated: 
```
		# Convert mm to cm 
		AvgDist = AvgDist / 10 
		event = 'Average Distance Measured: ' + str(AvgDist) + 'cm' # <=== Insert this line
		DebugWrite(event) # <=== Insert this line
 ```

**Reply by nebhead** [reply]
I don't know what I did, but it doesn't seem to be failing anymore, and it's getting reasonably accurate readings.  I mean, I added the debug code above, and it's all working.  Hmm... I may just construct a longer cable and try it on the real thing.

**Reply by faisalhusain007** [reply]
I have been playing around for my setup I have to angle the sensor slightly (crudely using a toothpick) but the measurements are consistent and fairly accurate.
![TOF test](https://user-images.githubusercontent.com/36872476/105570312-83e73a00-5d83-11eb-88ed-746d1e6b971d.jpg)

**Reply by faisalhusain007** [reply]
I am testing a 1.1M cable so far it still appears to detect on the smbus and works will continue testing, I am not using shielded cables just some spare LED lighting ribbon type cable.

**Reply by nebhead** [reply]
> I am testing a 1.1M cable so far it still appears to detect on the smbus and works will continue testing, I am not using shielded cables just some spare LED lighting ribbon type cable.

Oh excellent!  Great to know that it still works.  I need to take some time to get my "new" setup onto my grill so that I can try this out on the real thing.  Or maybe I'll take my bench setup outside. :)  Didn't have anytime to play with this yesterday - family time.  But will try to get to it this afternoon.

**Reply by faisalhusain007** [reply]
Yep family time is really important. All my kids are away studying, my sons in his final year of Uni, working on machine learning as his research and final year assignment. Interestingly he using python too.

**Reply by nebhead** [reply]
That's wonderful!  I kind of miss school sometimes - like perhaps I didn't value it as much when I was younger.  I would definitely like to learn more about machine learning someday.

**Comment by NZGriller** [comment]
@nebhead firstly, thank you for sharing a such an amazing project. I have 2 questions, Is it possible to use a thermocouple amplifier such as the Adafruit MCP9600 or AD8495 with PiFire? I'm working through the idea of installing PiFire into a brick pizza oven, which I assume will exceed the temperature limits of most RTD probes.
And lastly, is the ILI9341 display 'plug and play', or would I need to reconfigure the display coding?
My experience with Raspberry Pi projects is limited 1x HeaterMeter (in which I relied heavily on the instructional videos), so I'm a total noob with no coding experience.

**Reply by nebhead** [reply]
Hey @NZGriller!  Welcome to the discussions.  Very cool idea to extend this to a pizza oven!  Yes, it's entirely possible to extend PiFire to use a thermocouple amplifier such as those.  However, it may take a little bit of coding experience, and perhaps installing a python library to support one of those devices.  It's a great way to get your feet wet and learn!  

The way the PiFire works, is that it sets up an object for the ADC (basically anything that is fetching temperature).  PiFire imports the module adc_ads1115.py currently, which provides some basic functions: 

`SetProfiles()` - Selects the profiles for the different probes so they can be used in the calculations.
`ReadAllPorts()` - Reads from Grill Probe, Probe 1, and Probe 2 ports
`adctotemp()` - Returns temperature and resistance values for the probe selected 

You can basically copy this file, and rename it to something like `adc_mcp9600.py` or `adc_ad8495.py`, then make modifications to work with those devices.  Likely you'll need to install and import the right python libraries to support your code.  For example: `sudo pip3 install mcp9600`  (for the MCP9600)  (https://github.com/pimoroni/mcp9600-python) 

Once this is all good to go, you'll then need to make sure that the main PiFire control application loads your new module instead of the ads1115 module.  First you'll need to change the settings.json.  I have an relatively easy way to do this from the command line: 

`python3 settings.py -a mcp9600`

Then you'll need to modify the `control.py` application add some lines in the section that looks like the below: 

```
if(settings['modules']['adc'] == 'ads1115'):
	from adc_ads1115 import ReadADC # Library for reading the ADC device
elif(settings['modules']['adc'] == 'mcp9600'):  # Add this line here for your code
	from adc_mcp9600 import ReadADC    # Library for reading the mcp9600 device
else: 
	from adc_prototype import ReadADC # Simulated Library for reading the ADC device
```

That should be basically the main stuff you'll need to know.  Obviously, you'll probably run into issues, so I would recommend debugging at the command line (make sure you tell supervisor to stop trying to run control.py) by running control.py manually.  

As for the ili9341, it should be very easy to setup with the latest additions to the installation script.  Now you can just select the display type from the installation menu while walking through the install process and it should install the right library to get things going.  I personally just got my ili9341 screen installed on my grill and I think it looks pretty nice.  :) 

And lastly, I wanted to ask if your username implies you are from New Zealand?  That's actually the country I was born in so there is a special place in my heart for NZ.  

Just let me know if you have any more questions!  Good luck.

**Reply by NZGriller** [reply]
@nebhead thanks for the info. I think you're right that I need to order parts and jump in, and see what happens from there. Yes I'm down in little New Zealand. I think I will have more questions with my lack of experience, thanks again.

**Reply by nebhead** [reply]
Outstanding!  Feel free to ask questions here and I'll try to jump in and help when I can.  

As for your Pizza oven, is it based on wood pellets?

**Reply by NZGriller** [reply]
Thanks :)
I'm making a pizza oven from a kit set, and adding an auger/motor from Green Mountain Grills, with a custom pellet hopper and burn pot. So it will primarily be pallet fed.

**Comment by cptcave** [comment]
I really love this project and I'm about to get my first Smoker from Pit Boss but I want PID so this project seems a great fit as a replacement for the oem controller for that grill.  

I watched a video by youtuber MrBigKid https://youtu.be/yDnBvptAdJk.  He compared the Masterbuilt Gravity Charcoal grill and the pellet grill I'm looking at buying. He mentioned that one good feature on the Masterbuilt is a door switch that when triggered would stop the system from trying to compensate for the drastic temp change when the door opens.  He complained that the pellet grill seems to try and overcompensate for the temp change by dumping pellets once the door was opened.  Could a similar switch be included that would pause oven temp change until a certain time after the door closes?  I haven't had a chance to look through all of the features but wondered if you had experienced this issue.

I'm a little out of practice with programming so this project seems great for my skills.  BTW nebhead you mentioned you might would have some printed circuit boards available.  Do you have anymore of those boards and how might one go about getting one from you? Thanks and this is a Great Project!

**Reply by nebhead** [reply]
@cptcave - Welcome!  I think this idea of adding a switch is a very good idea!  I do have a few PCBS left, and if you want to send me an e-mail at _redacted_ with your address, I'll pop one in the mail to you.  I'll have to take a look at the video, but I'm curious how to get a switch attached to the lid without melting everything.  :)

**Reply by cptcave** [reply]
@nebhead -Thanks for responding and for the email.  At first thought, it would be no problem to find a heat resistant switch in the google world we live in, and they do exist.  I just can't seem to find a low volt option.  I imagine that the Masterbuilt switch may be low volt and I've looked at what one of those would cost.  It's about $50 and nobody has them.  Mainly because apparently they tend to fail (I guess a 700 degree grill doesn't do well against many electronics).  MB is supposedly sending updated versions with new grills and there isn't a part supply for them yet.  Every other oven door switch is in the ~$30+ range and are all 120/240v deals.  I am not sure if a high voltage switch would work to close a circuit that would normally be served by a 3.3v rated device or even be stable.  Who knew on/off switches would be so complicated?   I may not be looking in the right place or even asking the right question.  I'm even considering a magnetic switch or another optical sensor though both of those options seem unlikely with the heat involved.  Perhaps a long spring loaded actuator with a cheap switch away from the heat?  

I was also thinking that, as the door of the grill should be left open during startup, a system like this could be used to warn if the door isn't open during startup. You know, for those who can be absent minded, and their SOs who might like grilling as well.


...
I did find a tilt sensor that has a 100c temp rating for about $20.  I'm not sure what the temp of a grill lid can get but that might be going in the right direction.

**Reply by nebhead** [reply]
So, one rather low-tech solution might be to just attach some copper tape to the lid and to the body, which come in contact when the lid is closed.  Solder some wire to that and I suppose that would be enough.  I mean, a switch is nothing more than some contacts.  However, it would look rather messy, and might misbehave if it gets dirty.  But it's a thought.

**Reply by NZGriller** [reply]
HeaterMeter uses a button to manually enter Lid Open mode for a preset time, which stops the fan revving up. It also has an auto Lid Open mode that is triggered when the pit temp drops by a preset amount. Thinking aloud, maybe the PiFire version could "freeze" the current PID settings for a preset time to prevent the pot burning out if the lid is open for an extended period.
